python/nwm_client_new/src/hydrotools/nwm_client_new/ParquetStore.py

```diff
@@ -0,0 +1,144 @@
+"""
+==============================
+Parquet-backed DataFrame Store
+==============================
+Store dask dataframes using a key-value interface.
+
+Classes
+---------
+ - ParquetStore
+
+"""
+
+import dask.dataframe as dd
+from typing import Union, Iterator
+from pathlib import Path
+from collections.abc import MutableMapping
+
+class ParquetStore(MutableMapping):
```

Comment:
having talked to @jarq6c about this, the the issue comes to if `dask`'s io is read write safe. I don't think ill have the time today to throw together a test to try and verify this potential behavior, but when I do, ill add it here.

Reply:
I was able to produce IO race using `dask` and `parquet`. I verified this using `dask`'s `pyarrow` parquet engine. 

To do this, I spawned several dask futures and had the futures write different data to the same `parquet` store. For simplicity, some of the futures tried to write a dataframe of 0's and other futures tried to write 1's. All writers tried to write dataframe of the same length (nrows=100) and with the same number of partitions (nparts=100). Then I read from the parquet store and tried to assert that all the values are the same.

What I found is that sometime the store contains all of the same values after the read and other times it has a mixture of 1's and 0's indicating contention. Ive included the code below for anyone who would like to reproduce this themselves. Note, you may need to run it several times to get an assertion error.

Luckily, dask offers [some synchronization primitives](https://docs.dask.org/en/stable/futures.html#locks) that we should be able to use in this module to avoid read / write contention. Unfortunately, I didn't see a multiple reader or single writer lock (RWLock), however we should be able to accomplish something like this using Semaphores instead of a single lock. 

<details>
<summary>code
</summary>

```python
import time
import pandas as pd
import numpy as np
import dask.dataframe as dd
from dask.distributed import Client

# objective: verify if it is possible to create read write contention using dask

SAMPLE_SIZE = 100
PARTITIONS = 100
PARQUET_FILE = "data/df.parquet"


def fake_data_zeros() -> dd.DataFrame:
    return dd.from_pandas(
        pd.DataFrame({"values": np.zeros(SAMPLE_SIZE)}), npartitions=PARTITIONS
    )


def fake_data_ones() -> dd.DataFrame:
    return dd.from_pandas(
        pd.DataFrame({"values": np.ones(SAMPLE_SIZE)}), npartitions=PARTITIONS
    )


def read_parquet() -> dd.DataFrame:
    return dd.read_parquet(PARQUET_FILE)


def write(ddf: dd.DataFrame):
    return ddf.to_parquet(PARQUET_FILE)


if __name__ == "__main__":
    # this gist is, we want `to_parquet` and `read_parquet` to run at same time
    client = Client()

    zeros = client.submit(fake_data_zeros)
    ones = client.submit(fake_data_ones)

    for i in range(20):
        writer = client.submit(write, zeros)
        writer2 = client.submit(write, ones)

    # block
    writer.result()
    time.sleep(5)

    reader = client.submit(read_parquet)
    df = reader.result()
    value = df.head().iloc[0, 0]

    assert df.loc[df["values"] == value].all().compute().iat[0] == True

```

</details>

Reply:
Use of lock seems to avoid a mixture of 1's and 0's in the same file. However, which one writes last is still random. Updated code is below.

<details>
<summary>code
</summary>

```python
import time
import pandas as pd
import numpy as np
import dask.dataframe as dd
from dask.distributed import Client, Lock

# objective: verify if it is possible to create read write contention using dask

SAMPLE_SIZE = 100
PARTITIONS = 10
PARQUET_FILE = "data/df.parquet"


def fake_data_zeros() -> dd.DataFrame:
    return dd.from_pandas(
        pd.DataFrame({"values": np.zeros(SAMPLE_SIZE)}), npartitions=PARTITIONS
    )


def fake_data_ones() -> dd.DataFrame:
    return dd.from_pandas(
        pd.DataFrame({"values": np.ones(SAMPLE_SIZE)}), npartitions=PARTITIONS
    )


def read_parquet(lock) -> dd.DataFrame:
    with lock:
        df = dd.read_parquet(PARQUET_FILE)
    return df


def write(ddf: dd.DataFrame, lock):
    with lock:
        out = ddf.to_parquet(PARQUET_FILE)
    return out


if __name__ == "__main__":
    # this gist is, we want `to_parquet` and `read_parquet` to run at same time
    client = Client()

    lock = Lock()

    zeros = client.submit(fake_data_zeros)
    ones = client.submit(fake_data_ones)

    for i in range(2):
        writer = client.submit(write, zeros, lock)
        writer2 = client.submit(write, ones, lock)

    # block
    writer.result()
    time.sleep(5)

    reader = client.submit(read_parquet, lock)
    df = reader.result()
    value = df.head().iloc[0, 0]

    assert (df["values"] == value).all().compute() == True


```

</details>

Reply:
Yeah, fair point. Do we have a use case where this module would be used where last writer wins semantics are problematic? 🤔

Reply:
This is probably overkill for this use case, but it might come in helpful in the future. One drawback of using a lock here is that you can only have a single reader or single writer. Ideally, you should be able to have as many readers as you want, as long as there are is not a writer. Likewise, the only time you should be able to write is when there are no readers. The below code implements an RWLock using `dask`'s synchronization primitives.

The main benefit of using an RWLock in this circumstance is multiple readers can do something with the data guarded by the lock. 

The two main drawbacks to using the below code are:

1. Added implementation complexity.
2. Readers have to grab the twice as many times as they do now.


<details>
<summary>RWLock Code</summary>

```python
from dataclasses import dataclass, field
from dask.distributed import Lock, Event, Variable


@dataclass
class RWLock:
    _lock: Lock = field(default_factory=Lock)
    _rc_name: str = "rwlock"
    _rc: Variable = field(default_factory=lambda: Variable(RWLock._rc_name))
    _convar: Event = field(default_factory=Event)

    def __post_init__(self):
        # there are no readers or writers at this point, so the reader reference is 0 and the convar
        # is set, meaning it is safe to read or write.
        self._rc.set(0)
        self._convar.set()

    def read(self):
        """
        Example:

        rwlock = RWLock()

        with rwlock.read():
            # read things
        """
        return _RLock(lock=self._lock, rc=self._rc, convar=self._convar)

    def write(self):
        """
        Example:

        rwlock = RWLock()

        with rwlock.write():
            # write things
        """
        return _WLock(lock=self._lock, rc=self._rc, convar=self._convar)


@dataclass
class _Lock:
    lock: Lock
    rc: Variable
    convar: Event


@dataclass
class _RLock(_Lock):
    """readers acquire lock 2 times as much as writers. however, the time they hold the lock should
    generally be shorter than writers.
    """

    def __enter__(self):
        self.lock.acquire()

        # block any potential writers
        if self.convar.is_set():
            self.convar.clear()

        # increment reader reference count by 1
        try:
            n_readers = self.rc.get()
            n_readers += 1
            self.rc.set(n_readers)
        finally:
            self.lock.release()

    def __exit__(self, exc_type, exc_value, traceback):
        self.lock.acquire()

        # decrement reader reference count by 1
        try:
            n_readers = self.rc.get()
            n_readers -= 1
            self.rc.set(n_readers)

            # no more readers, so tell writers it is okay to write
            if n_readers == 0:
                self.convar.set()
        finally:
            self.lock.release()


@dataclass
class _WLock(_Lock):
    def __enter__(self):
        self.lock.acquire()

        # writer acquires lock but convar is not set. meaning it is not safe to write.
        # release lock and wait for convar to be set
        # re-acquire lock, then check if convar has been set. this indicates that all readers are
        # done reading and no readers have the lock, so it is safe to write.
        #
        # if a new reader acquired the lock right after all readers were done, it is possible for
        # the convar to have been set before re-acquiring the lock.
        if not self.convar.is_set():
            while True:
                self.lock.release()
                # block until readers signal they are done
                self.convar.wait()

                # multiple writers will fight over acquiring this lock
                self.lock.acquire()
                if self.convar.is_set():
                    # safe to proceed. no readers have lock and convar is set
                    break
                # implicit continue if convar is now not set

        # implicit convar must be set, so safe to write

    def __exit__(self, exc_type, exc_value, traceback):
        self.lock.release()
```


</details>

Reply:
Given the potential complexities involved, I've decided to add a note to the documentation that this is a collection of convenience methods for a common data access pattern. Users should probably avoid parallel access to `ParquetStore` in favor of using `dask` methods directly. The source parquet file used by `ParquetStore` is just a normal parquet file, after all.